Extension codestart guide (and javadoc for QuarkusCodestartTest) (#20304)
quarkusio/quarkus: docs/src/main/asciidoc/extension-codestart.adoc

////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/main/docs/src/main/asciidoc
////
= Extension codestart
include::./attributes.adoc[]

This guide explains how to create and configure a Quarkus Codestart for an extension.

== Description

"Extension Codestarts" is the name we gave to our Quarkus extension quickstart code generation system. It aims to provide a personalized getting started experience with Quarkus.
A Quarkus extension is able to provide one or more well defined codestarts which will contain the necessary resources and code required to start using that particular extension.

> You can play with the extension codestarts in the Quarkus tooling:

Reviewer: ```suggestion
You can apply with the extension codestarts in the Quarkus tooling:
```
I don't insist on it, I would suggest not to use "play" but something else.

* https://code.quarkus.io[code.quarkus.io, window="_blank"] (find the extensions tagged with [code])
* The Quarkus Maven plugin:
+
[source,bash]
----

> mvn io.quarkus:quarkus-maven-plugin:create

Reviewer: ```suggestion
mvn io.quarkus.platform:quarkus-maven-plugin:create
```

----

* The Quarkus CLI:
+
[source,bash]
----
quarkus create app
----

== How it works

> There are two kinds of codestarts contributing to the generation of a project, the kind where we want to have only one for a project (called Base) and the rest (called Extra).

Reviewer: I find this sentence a bit confusing but also struggle to suggest something better.

Author: I rewrited it

Base:

* project: The project skeleton (e.g. a Quarkus project)
* buildtool: The build tool (e.g. Maven, Gradle, Gradle with Kotlin DSL)
* language: The coding language (e.g. Java, Kotlin, Scala)
* config: The config type (e.g. yaml, properties)

Extra:

> * tooling: Anything that can be added to improve the project (e.g. dockerfiles)

Reviewer: Would `config` be a better choice here than `tooling`?
```suggestion
* configuration: Anything that can be added to enhance the project configuration (e.g. dockerfiles)
```

Author: `config` is already a base (app-config)

* code: Any Quarkus extension can provide starter code. The user can decide to activate it or not.

Each codestart consists of:

. A codestart unique name, ie `my-codestart`
. A directory for the codestart files, ie `my-codestart/`
. A `codestart.yml` file

> . Optionally some templates which following a common structure and naming conventions

Reviewer: ```suggestion
. Optionally some templates that are following a common structure and naming conventions
```

== Where are the Quarkus Extension Codestarts located

- In Quarkus core repo, the extension codestarts are all in the same https://github.com/quarkusio/quarkus/tree/main/devtools/project-core-extension-codestarts/src/main/resources/codestarts/quarkus/extension-codestarts[module, window="_blank"].

- RESTEasy, RESTEasy Reactive and Spring Web extension codestarts are part of https://github.com/quarkusio/quarkus/tree/main/independent-projects/tools/base-codestarts/src/main/resources/codestarts/quarkus/extension-codestarts[the base codestarts, window="_blank"].

> - For non-core extensions, the codestart is located in the runtime module (with special instruction in the pom.xml to generate a separated codestart artifact).

Reviewer: ```suggestion
- For other extensions, the codestart will typically be located in the runtime module (with special instruction in the `pom.xml` to generate a separate codestart artifact).
```

== Base codestarts

The https://github.com/quarkusio/quarkus/tree/main/independent-projects/tools/base-codestarts/src/main/resources/codestarts/quarkus[base codestarts, window="_blank"] contains templates to create project, buildtool, languages, config & tooling files.

== Writing an Extension Codestart

> As we discussed before, the base project files (pom.xml, dockerfiles, ...) are already generated by some existing base codestarts. Thanks to this, we can only focus on the important, the starter code for the extension.

Reviewer: ```suggestion
As was mentioned previously, the base project files (pom.xml, dockerfiles, ...) are already generated by base codestarts provided by the Quarkus core. Thanks to this, we can only focus on the important - the starter code for the extension.
```

> The codestart should not have any business code, instead, it should contain some stub data/hello world to compile. The idea is to bring code that is the starting point to everyone using the extension.

Reviewer: ```suggestion
The codestart should not include any business logic, instead, it should contain some stub data/hello world to compile. The idea is to bring code that is the starting point to everyone using the extension.
```

== Writing an Extension Codestart in Quarkus Core

> - Duplicate an existing https://github.com/quarkusio/quarkus/tree/main/devtools/project-core-extension-codestarts/src/main/resources/codestarts/quarkus/extension-codestarts[Quarkus core extension codestarts, window="_blank"]. `resteasy-qute-codestart` is a good base when the code will expose a web resource, in other cases, `config-yaml-codestart` is a better base. More info on the <<directory-structure>>.

Reviewer: ```suggestion
- Copy one of the existing https://github.com/quarkusio/quarkus/tree/main/devtools/project-core-extension-codestarts/src/main/resources/codestarts/quarkus/extension-codestarts[Quarkus core extension codestarts, window="_blank"]. If the code needs to expose a web resource, `resteasy-qute-codestart` could be a good base Otherwise, `config-yaml-codestart` could be a better starting point. More info on the <<directory-structure>>.
```

- Edit the <<codestart-yml>>:

- Create the extension binding in the extension metadata (https://github.com/quarkusio/quarkus/blob/main/extensions/config-yaml/runtime/src/main/resources/META-INF/quarkus-extension.yaml#L12-L17[example, window="_blank"]). *Thanks to this, the codestart is added when the user selects the extension*

- Add the readme <<readme-md>> section template.

> - Add the code in the language folder (it is recommended to at least provide java and kotlin). *You have to use org.acme as package name: <<org-acme-package>>*. It is possible to use <<qute-templates>> if needed.

Reviewer: ```suggestion
- Add the code in the language folder (it is recommended to at least provide java and kotlin). *You have to use `org.acme` as the package name: <<org-acme-package>>*. It is possible to use <<qute-templates>> if needed.
```

- Optionally, Add the `index.html` section template (<<index-html>>).

- Optionally, add some resources (`./base` directory if they are non language specific)

- Optionally, add the <<app-config>>.

- Create an <<integration-test>> *You can use it to get your codestart generated as a project*.

> You can also use it for dev using the `buildAllProjectsForLocalUse` (It's disabled by default because already covered by `QuarkusCodestartBuildIT`).

Reviewer: ```suggestion
You can also use it for dev using the `buildAllProjectsForLocalUse` (Disabled by default because it's already covered by `QuarkusCodestartBuildIT`).
```

== Writing an Extension Codestart in the Quarkiverse or standalone

> For non-core extensions, the codestart is located in the runtime module (with special instruction in the pom.xml to generate a separated codestart artifact). https://github.com/ia3andy/aloha-code/[Here, window="_blank"] is an example extension with a codestart & its tests

Reviewer: ```suggestion
For extensions hosted outside of the Quarkus core[https://github.com/quarkusio/quarkus] repository, codestarts will typically be located in the runtime module (with special instruction in the `pom.xml` to generate a separate codestart artifact). https://github.com/ia3andy/aloha-code/[Here, window="_blank"] is an example extension with a codestart & its tests
```

== Specific topics

[#org-acme-package]
=== Dynamic package name generation from org.acme

> You have to use `org.acme` as package name in your extension codestart sources. In the generated project, the user specified package will be used (and auto-replace `org.acme`).

Reviewer: ```suggestion
You have to use `org.acme` as the package name in your extension codestart sources. In the generated project, the user specified package will be used (and auto-replace `org.acme`).
```

> In all the source files (.java, .kt, .scala), it will be auto-replaced, you don't have to care. The package directory also gets auto-replace. If for some reason, another type of file needs the user package name, then you should use a <<qute-templates>> for it and `{project.package-name}` data placeholder (https://github.com/quarkusio/quarkus/blob/main/devtools/project-core-extension-codestarts/src/main/resources/codestarts/quarkus/extension-codestarts/grpc-codestart/base/src/main/proto/hello.tpl.qute.proto#L4[find an example in the grpc proto file, window="_blank"]).

Reviewer: ```suggestion
It will be auto-replaced in all the source files (.java, .kt, .scala). The package directory will also be automatically adjusted. If for some reason, another type of file needs the user package name then you should use a <<qute-templates>> for it and `{project.package-name}` data placeholder (https://github.com/quarkusio/quarkus/blob/main/devtools/project-core-extension-codestarts/src/main/resources/codestarts/quarkus/extension-codestarts/grpc-codestart/base/src/main/proto/hello.tpl.qute.proto#L4[find an example in the grpc proto file, window="_blank"]).
```

metadata:
  title: RESTEasy JAX-RS example
  description: Rest is easy peasy with this Hello World RESTEasy resource.
  related-guide-section: https://quarkus.io/guides/getting-started#the-jax-rs-resources
  # the path is optional and used by the generated index.html if present
  path: /some-path
language:
  base:
    # Specify the extension and possibly other required dependencies
    dependencies:
      - io.quarkus:quarkus-resteasy
    # And maybe test dependencies?
    test-dependencies:
      - io.rest-assured:rest-assured
----

[#directory-structure]
=== Directory Structure

NOTE: `codestart.yml` is the only required file.

* `codestart.yml` must be at the root of the codestart
* `./base` contains all the files that will be processed independently of the specified language
* `./[java/kotlin/scala]` contains all the files that will be processed only if the specified language has been selected (overriding base)

=== Naming Convention for files

* `.tpl.qute` will be processed with Qute and can use data (`.tpl.qute` will be removed from the output file name).

> * some files have a specific processing (`readme.md`, `src/test/resources/application.yml`, `src/main/resources/META-INF/resources/index.html`)

Reviewer: ```suggestion
* certain common files, such as `readme.md`, `src/test/resources/application.yml`, `src/main/resources/META-INF/resources/index.html` are generated from the collected fragments found in the selected codestarts for the project
```

An extension is available to help test extension codestarts `QuarkusCodestartTest`. It provides a way to test:

- the generated project content (with immutable mocked data) using snapshot testing
- the generated project build/run (with real data) with helpers to run the build

NOTE: Before all tests, the extension will generate Quarkus projects in the specified languages with the given codestart, with mocked data and with real data. You can find those generated project in `target/quarkus-codestart-test`. You can open the `real-data` ones in your IDE or play with them using the terminal. *Running those tests is the easiest way to iterate on your extension codestart dev.*

The extension provides helpers to test that the projects build `buildAllProjects` or just a specific language project `buildProject(Language language)`. It also provides helpers to test the content with <<snapshot-testing>>.

The https://github.com/quarkusio/quarkus/blob/main/integration-tests/devtools/src/test/java/io/quarkus/devtools/codestarts/quarkus/ConfigYamlCodestartTest.java[ ConfigYamlCodestartTest, window="_blank"] is a good example in Quarkus core.

[#snapshot-testing]
==== Snapshot testing

> Snapshot testing is way to make sure the generation output is always "as expected" across commits. For this the generated content for one commit needs to be immutable and determinist (this is the reason for using mocked data). For this we auto-generate and commit snapshots, they become the reference of the output. When the templates change, we also commit the induced snapshots changes. This way, in the review, we can make sure the modification is working "as expected".

Reviewer: ```suggestion
Snapshot testing is a way to make sure the content generated by a test doesn't change from one revision to another, i.e. between commits. That means, the generated content for each commit needs to be immutable and deterministic (this is the reason for using mocked data). To be able to perform such checks, we auto-generate snapshots of the generated content  and commit them as the references of the expected output for subsequent test runs. When the templates change, we also commit the induced snapshots changes. This way, during the review, we can make sure the applied code changes have the expected effects on the generated output.
```
